### PR TITLE
[ko] guide/create_op

### PR DESCRIPTION
**Pull request by lamberta**
MIgrate https://github.com/tensorflow/docs/pull/1257 by @wckim
CLA signed in that PR

### CODE REVIEW

**Comment by jaketae** [comment]
Hi @lamberta 

@Choiuijin1125 has already left an extensive review of this submission on the original PR. Is there any way we can migrate the comments to here as well? Thank you.

**Comment by lamberta** [comment]
Thanks for looking. Unfortunately, there doesn't seem to be a way to migrate comments over. (It was kind of hacky just migrating the original commits to a different repo.)

Since the pull request author has changed (to me), if you make a suggested edit using GitHub's UI, I can go ahead and merge them. Since there's a backlog, I'm less concerned about perfect quality---the community can still submit corrective pull requests once this is merged.

**Comment by jaketae** [comment]
@lamberta Understood. I'll try to make suggested edits based on the comments available on the original PR thread.

File: site/ko/guide/create_op.md [hunk 1]
```diff
@@ -0,0 +1,1347 @@
...
+    편의상 파이썬에서 테스트를 하지만 C++에서 테스트를 할 수 있습니다.
+    그래디언트를 정의했다면, 파이썬 `tf.test.compute_gradient_error`를 활용해 검증할 수 있습니다.
+    렐루(Relu)와 유사한 연산자의 정방향 함수와 그들의 그래디언트를 테스트하는 예인
+    [`relu_op_test.py`](https://www.tensorflow.org/code/tensorflow/python/kernel_tests/relu_op_test.py)를 참고하세요.
+
+### 선행조건
+
+* C++에 익숙
+* [텐서플로 실행파일](../../install)이 설치되어 있거나
+  [텐서플로 소스를 다운로드](../../install/source.md)해서 그 소스를 설치할 수 있어야 함
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit " 그 소스를 설치할 수 있어야 함 " to " 그 소스를 설치할 수 있어야 합니다."

File: site/ko/guide/create_op.md [hunk 2]
```diff
@@ -0,0 +1,1347 @@
...
+4.  연산을 위한 그래디언트를 계산할 함수를 작성하세요 (선택).
+5.  연산을 테스트하세요.
+    편의상 파이썬에서 테스트를 하지만 C++에서 테스트를 할 수 있습니다.
+    그래디언트를 정의했다면, 파이썬 `tf.test.compute_gradient_error`를 활용해 검증할 수 있습니다.
+    렐루(Relu)와 유사한 연산자의 정방향 함수와 그들의 그래디언트를 테스트하는 예인
+    [`relu_op_test.py`](https://www.tensorflow.org/code/tensorflow/python/kernel_tests/relu_op_test.py)를 참고하세요.
+
+### 선행조건
+
+* C++에 익숙
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "C++에 익숙" to "C++ 언어를 익숙하게 다룰 줄 알아야 합니다."

File: site/ko/guide/create_op.md [hunk 3]
```diff
@@ -0,0 +1,1347 @@
...
+### GPU 커널
+
+GPU 커널 구현은 두 부분으로 구성되어 있습니다: OpKernel과 CUDA 커널 및 관련 실행 코드
+
+OpKernel 구현에서 입력값 검사나 출력을 할당하는 부분은
+CPU 커널과 GPU 커널에서 공통적으로 활용되기도 합니다.
+이런 경우 제안하는 구현 방식은:
+
+1. 기기와 텐서 변수형을 이용한 템플릿 클래스로 OpKernel을 정의합니다.
+2. 실제 결과를 계산하기 위해 Compute 함수는 템플릿 인자로 호출합니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Add inline code markdown to the word "Compute" by changing it to `Compute`

File: site/ko/guide/create_op.md [hunk 4]
```diff
@@ -0,0 +1,1347 @@
...
+struct ExampleFunctor<CPUDevice, T> {
+  void operator()(const CPUDevice& d, int size, const T* in, T* out) {
+    for (int i = 0; i < size; ++i) {
+      out[i] = 2 * in[i];
+    }
+  }
+};
+
+// OpKernel 정의
+// 템플릿 매개변수 <T>은 텐서 형태
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "<T>은" to "<T>는"

File: site/ko/guide/create_op.md [hunk 5]
```diff
@@ -0,0 +1,1347 @@
...
+with tf.Session(''):
+  zero_out_module.zero_out([[1, 2], [3, 4]]).eval()
+
+# 출력
+array([[1, 0], [0, 0]], dtype=int32)
+```
+
+생성된 함수는 스네이크_케이스(snake\_case)형태 이름으로 주어집니다
+([PEP8](https://www.python.org/dev/peps/pep-0008/)를 준수하기 위해).
+그래서 작성한 연산 이름이 C++ 파일에서 `ZeroOut`이라면 파이썬 함수는 `zero_out`라는 이름으로 호출됩니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "`zero_out`라는" to "`zero_out`이라는"

File: site/ko/guide/create_op.md [hunk 6]
```diff
@@ -0,0 +1,1347 @@
...
+    *   [입력과 출력](#입력과-출력)
+    *   [하위 호환성](#하위-호환성)
+*   [GPU 지원](#GPU-지원)
+    *   [GPU를 위한 커널 컴파일하기](#GPU를-위한-커널-컴파일하기)
+*   [파이썬에서 그래디언트 구현하기](#파이썬에서-그래디언트-구현하기)
+*   [C++에서 형태 함수](#c에서-형태-함수)
+
+### 조건부 검사와 검증
+
+위의 예제는 연산을 어떤 형태 텐서에도 활용 가능하다고 가정했습니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "어떤 형태 텐서" to "어떤 형태의 텐서"

File: site/ko/guide/create_op.md [hunk 7]
```diff
@@ -0,0 +1,1347 @@
...
+
+추가된 부분은 입력값이 벡터임을 확인하고 벡터가 아니라면 `InvalidArgument`인 상태를 반환합니다.
+추가된 [`OP_REQUIRES` 매크로][validation-macros]는 입력값으로 3개를 받습니다:
+
+*   `컨텍스트`, `SetStatus()` 메소드를 위해서 `OpKernelContext` 혹은 `OpKernelConstruction`의 포인터
+    ([`tensorflow/core/framework/op_kernel.h`](https://www.tensorflow.org/code/tensorflow/core/framework/op_kernel.h) 참조)
+*   상태. 예를 들어, 텐서 형태를 확인하는 함수들은
+    [`tensorflow/core/framework/tensor_shape.h`](https://www.tensorflow.org/code/tensorflow/core/framework/tensor_shape.h)에서 확인할 수 있습니다.
+*   에러, `Status` 객체로 표현됩니다,
+    [`tensorflow/core/lib/core/status.h`](https://www.tensorflow.org/code/tensorflow/core/lib/core/status.h) 참조하세요.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "...) 참조하세요." to "...)를 참조하세요."

File: site/ko/guide/create_op.md [hunk 8]
```diff
@@ -0,0 +1,1347 @@
...
+#### 속성
+
+연산은 그래프에 추가될 때 설정되는 값인 속성을 가질 수 있습니다.
+속성은 연산을 설정하는 데 사용되는데, 커널 구현에서 활용되거나
+연산 등록에서 입력과 출력의 자료형으로 사용될 수 있습니다.
+입력값은 다양하기 때문에 가능하다면 속성 대신에 입력값을 사용하는 것을 선호합니다.
+속성은 고정값이고 그래프가 생성되는 시점에 정의되어져야 하기 때문입니다.
+반대로 입력은 입력할 때마다 매번 변경될 수 있고 사용자에 의해서 설정할 수 있는 동적으로 변경되는 텐서입니다.
+속성은 입력을 활용할 수 없는 경우에 사용됩니다:
+서명(입력과 출력의 개수와 자료형)에 영향을 미치거나 매번 변경될 수 없는 어떤 구성.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit the entire line to "속성은 서명(입력과 출력의 개수와 자료형)에 영향을 미치거나 매번 변경될 수 없는 구성입니다."

File: site/ko/guide/create_op.md [hunk 9]
```diff
@@ -0,0 +1,1347 @@
...
+* `{<type1>, <type2>}`: 값은 `type` 자료형이고 `tf.DType`에서 지원하는 `<type1>` 또는 `<type2>` 중에 하나여야 합니다.
+  속성 자료형에 `type`을 명시하지 않습니다. 그것은 `{...}`안의 자료형 리스트를 통해 암시됩니다.
+  예를 들어, 아래와 같은 경우에 속성 `t`의 자료형은 `int32`과 `float`, `bool` 중에 하나입니다:
+
+  ```c++
+  REGISTER_OP("RestrictedTypeExample")
+      .Attr("t: {int32, float, bool}");
+  ```
+
+* 일반적인 자료형에 대한 제약사항에 대한 몇 가지 단축형이 있습니다:
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "일반적인 자료형에 대한" to "일반적인 자료형이 가지는"

File: site/ko/guide/create_op.md [hunk 10]
```diff
@@ -0,0 +1,1347 @@
...
+* `list(<type>) >= <n>`: `<n>`보다 크거나 같은 길이를 가진 `<type>`의 리스트입니다.
+
+  예를 들어, 다음 연산 등록은 속성 `a`가 `int32` 혹은 `float`의 리스트이고 그 값이 최소한 3보다 큰 값임을 명시합니다:
+
+  ```c++
+  REGISTER_OP("TypeListExample")
+      .Attr("a: list({int32, float}) >= 3");
+  ```
+
+속성에 기본값을 할당하려면(이를 통해 생성된 코드에서 속성을 옵셔널(optional)으로 만들어줌),
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "만들어줌" to "만들어줍니다"

File: site/ko/guide/create_op.md [hunk 11]
```diff
@@ -0,0 +1,1347 @@
...
+```c++
+REGISTER_OP("MinimumLengthPolymorphicListExample")
+    .Attr("T: list(type) >= 3")
+    .Input("in: T")
+    .Output("out: T");
+```
+
+#### 입력과 출력
+
+위의 내용을 요약하자면, 다양한 입력과 출력을 대해서 연산을 등록할 수 있습니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "입력과 출력을" to "입력과 출력에"

File: site/ko/guide/create_op.md [hunk 12]
```diff
@@ -0,0 +1,1347 @@
...
+           .Attr("T: numerictype = DT_FLOAT");
+
+2. 속성에 대한 제약을 줄여서 안전하게 만들 수 있습니다.
+   예를 들어, `{int32, int64}`을 `{int32, int64, float}` 혹은 `type`으로 변경가능합니다.
+   또는, `{"apple", "orange"}`을 `{"apple", "banana", "orange"}` 혹은 `string`으로 변경할 수 있습니다.
+
+3. 리스트의 기본값을 변경 이전과 동일하게 맞춘다면,
+   단일 입력/출력을 리스트로된 입력/출력으로 변경할 수 있습니다.
+
+4. 기본값이 없다면, 새로운 입력/출력 리스트룰 추가할 수 있습니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "리스트룰" to "리스트를"

File: site/ko/guide/create_op.md [hunk 13]
```diff
@@ -0,0 +1,1347 @@
...
+
+```bash
+TF_CFLAGS=( $(python -c 'import tensorflow as tf; print(" ".join(tf.sysconfig.get_compile_flags()))') )
+TF_LFLAGS=( $(python -c 'import tensorflow as tf; print(" ".join(tf.sysconfig.get_link_flags()))') )
+g++ -std=c++11 -shared zero_out.cc -o zero_out.so -fPIC ${TF_CFLAGS[@]} ${TF_LFLAGS[@]} -O2
+```
+
+Max OS X에서 `.so`을 생성할 때 "-undefined dynamic_lookup"이라는 추가적인 플래그가 필요합니다.
+
+>   `gcc` 버젼이 `>=5`인 경우 주의사항:
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "버젼" to "버전"

File: site/ko/guide/create_op.md [hunk 14]
```diff
@@ -0,0 +1,1347 @@
...
+>    gcc는 버젼 `5`부터 새로운 C++ [ABI](https://gcc.gnu.org/gcc-5/changes.html#libstdcxx)를 사용합니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "버젼" to "버전"

File: site/ko/guide/create_op.md [hunk 15]
```diff
@@ -0,0 +1,1347 @@
...
+안전하거나 안전하지 않는 변경에 대한 전체 내용은
+[`tensorflow/core/framework/op_compatibility_test.cc`](https://www.tensorflow.org/code/tensorflow/core/framework/op_compatibility_test.cc)에서
+찾을 수 있습니다.
+만약에 하위 호환성을 확보하기 어렵다면, 그때는 새로운 이름과 의미를 갖는 새로운 연산을 만드세요.
+
+또한 이러한 변경을 `GraphDef` 호환성을 지키면서 유지할 수 있는 동안,
+생성되는 파이썬 코드는 이전 버전과 호환되지 않는 방식으로 변경될 수 있음을 주의하세요.
+새로운 선택형 파라미터를 추가하는 것을 제외하고 이전 버전의 특징을 유지함으로써,
+파이썬 API는 사용자가 파이썬 랩퍼를 신중하게 변경하여 호환성을 유지할 수 있습니다.
+일반적으로 호환되지 않는 변경은 텐서플로 메이저 버젼이 변경되고,
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "버젼" to "버전"

File: site/ko/guide/create_op.md [hunk 16]
```diff
@@ -0,0 +1,1347 @@
...
+[`GraphDef` 버전 의미](./versions.md#compatibility_of_graphs_and_checkpoints)를 준수해야 하는 경우에만 발생합니다.
+
+### GPU 지원
+
+OpKernel을 구현할 때 CPU를 지원하는 것과 GPU를 지원하는 것으로 다르게 구현을 할 수 있고,
+이를 [다른 자료형을 위한 커널 등록](#다형성)과 같이 처리할 수 있습니다.
+[`tensorflow/core/kernels/`](https://www.tensorflow.org/code/tensorflow/core/kernels/)는
+GPU를 지원하는 커널의 예입니다.
+어떤 커널은 `.cc`로 끝나는 파일이 CPU버젼이고 `_gpu.cu.cc`로 끝나는 파일이 GPU버젼이고
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "GPU버젼" to "GPU 버전"

File: site/ko/guide/create_op.md [hunk 17]
```diff
@@ -0,0 +1,1347 @@
...
+
+예를 들어, `tf.pad`는 GPU 커널을 위한 부분을 제외하고 [`tensorflow/core/kernels/pad_op.cc`][pad_op]에 모든 구현되어 있습니다.
+[`tensorflow/core/kernels/pad_op_gpu.cu.cc`](https://www.tensorflow.org/code/tensorflow/core/kernels/pad_op_gpu.cu.cc)에
+GPU커널에 관련된 부분이 있고, 공유하는 코드는 템플릿 클래스 형태로
+[`tensorflow/core/kernels/pad_op.h`](https://www.tensorflow.org/code/tensorflow/core/kernels/pad_op.h)에 정의되어 있습니다.
+코드를 이처럼 관리하는 이유는 2가지입니다:
+CPU를 위한 구현과 GPU를 위한 구현에 필요한 공통 코드를 공유할 수 있고
+GPU 컴파일러가 컴파일하는 GPU구현 부분만 다른 파일로 분리할 수 있습니다.
+
+추가적으로 알아야할 것은, `pad`의 GPU 커널 버젼이 사용되는 경우에도,
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "버젼" to "버전"

File: site/ko/guide/create_op.md [hunk 18]
```diff
@@ -0,0 +1,1347 @@
...
+
+```c++
+REGISTER_OP("ZeroOut")
+    .Input("to_zero: int32")
+    .Output("zeroed: int32")
+    .SetShapeFn(::tensorflow::shape_inference::UnchangedShape);
+```
+
+형태 함수는 입력을 형태를 제한할 수 있습니다.
+[`ZeroOut`의 입력을 벡터 형태로 제한한](#조건부-검사와-검증) 버젼의 경우, 형태 함수는 다음과 같을 것입니다:
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "버젼" to "버전"

File: site/ko/guide/create_op.md [hunk 19]
```diff
@@ -0,0 +1,1347 @@
...
+만약에 쿠다 라이브러리가 `/usr/local/lib64`에 설치되어 있지 않다면,
+위의 2번째 명령어(g++)에 라이브러리 경로를 명시적으로 추가할 수 있습니다.
+예를 들어, 쿠다가 `/usr/local/cuda-8.0`에 설치되어 있다면 `-L /usr/local/cuda-8.0/lib64/`를 추가하세요.
+
+>   일부 리눅스에서는, `nvcc` 컴파일 단계에서 추가적인 설정이 필요합니다.
+>   `nvcc` 명령어에서 `mwaitxintrin.h`때문에 생길 수 있는 에러를 피하기 위해 `-D_MWAITXINTRIN_H_INCLUDED`를 추가하세요.
+
+### 파이썬에서 그래디언트 구현하기
+
+연산 그래프가 주어지면, 텐서플로는 기존 연산에 대한 그래디언트를 표현하는 새로운 연산을 추가하기 위해서 자동미분(역전파) 사용합니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "자동미분(역전파)" to "자동미분(역전파)을"

File: site/ko/guide/create_op.md [hunk 20]
```diff
@@ -0,0 +1,1347 @@
...
+  예를 들어, 실수형 텐서 `x`와 정수 인덱스 `i`를 입력으로 받아들이는 연산인 경우
+  그래디언트 함수는 `return [x_grad, None]`일 것입니다.
+
+* 연산을 위해 의미가 있는 그래디언트가 없다면, 보통 어떤 그래디언트도 등록하지 않을 것이고,
+  연산의 그래디언트가 전혀 필요없는 한 그 것은 괜찮을 것입니다.
+  어떤 경우에, 연산은 잘 정의된 그래디언트를 가지고 있지 않지만 그래디언트의 계산에 관여될 수 있습니다.
+  그런 경우 자동적으로 0을 역전파하기 위해서 `ops.NotDifferentiable`를 사용할 수 있습니다.
+
+그 때 그래디언트 함수가 호출 될 때 텐서 그 자체가 아니라 연산 데이터 흐름 그래프만 사용할 수 있음을 유의하세요.
+그래서, 모든 계산은 그래프 실행시간에 실행되기 위해서 다른 텐스플로 연산을 사용해서 수행해야 합니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "텐스플로" to "텐서플로우"

File: site/ko/guide/create_op.md [hunk 21]
```diff
@@ -0,0 +1,1347 @@
...
+다음과 같이 사용됩니다:
+
+```c++
+REGISTER_OP("ZeroOut")
+    .Input("to_zero: int32")
+    .Output("zeroed: int32")
+    .SetShapeFn(::tensorflow::shape_inference::UnchangedShape);
+```
+
+형태 함수는 입력을 형태를 제한할 수 있습니다.
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "입력을" to "입력의"

File: site/ko/guide/create_op.md [hunk 22]
```diff
@@ -0,0 +1,1347 @@
...
+.SetShapeFn([](::tensorflow::shape_inference::InferenceContext* c) {
+    c->set_output(0, c->Matrix(c->Dim(c->input(0), 0), 3));
+    return Status::OK();
+});
+```
+
+만약 복잡한 형태 함수가 있다면,
+다양한 입력 형태 조합이 기대한 출력 형태 조합을 만들어 내는지 검증하기 위한 테스트를 추가하는 것을 고려해야 합니다.
+그러한 테스트를 어떻게 작성하는지에 대한 예는
+[core ops tests](https://www.tensorflow.org/code/tensorflow/core/ops/array_ops_test.cc)에서 볼 수 있습니다
```
Comment (jaketae) [inline site/ko/guide/create_op.md]: Edit "core ops tests" to "주요 연산 테스트"

**Comment by lamberta** [comment]
> I'm working under the assumption that you don't speak Korean, but correct me if I'm wrong

Good assumption 😀 But I do know "annyeonghaseyo"

When writing comments in a pull request, you can insert suggestions that the author can click to accept: https://til.hashrocket.com/posts/szgdi4at59-github-insert-a-suggestion-

**Comment by jaketae** [comment]
@lamberta Thanks for the heads up! Your Korean is up there, don't get me wrong, but something tells me that suggestions is going to work better than me writing down comments and you making manual fixes for each. I've changed my previous comments to suggestions, but this is obviously my first time trying so let me know if anything seems out of order. Thanks!

**Pull request closed without merging**